### DISCUSSION TITLE
embedded memtest86 app in pei phase

### DISCUSSION THREAD

**Discussion by yoshinoyatoko** [memtest86plus/memtest86plus#420] [General]
Hello,

I have an idea, but not sure it's possible or not.

Take UEFI BIOS as an example:

Usually init board dram during pei phase.

I plan to launch memory test immediately after dram init.

Is it possible to launch memtest86 app during UEFI BIOS's PEI phase?

I can include memtest86.efi as an pei driver into uefi bios bin.

**Comment by martinwhitaker** [comment]
Memtest86+ uses the EFI Boot Services to obtain the physical memory map and uses the EFI Graphics Output Protocol to locate and configure the video frame buffer. Are those services/protocols available in the PEI phase? I suspect not.

Also Memtest86+ overwrites the memory regions used by the UEFI BIOS, so once it has run there is no possibility of returning to the UEFI BIOS except by rebooting. So what do you gain by running it during the PEI phase?

**Comment by yoshinoyatoko** [comment]
Hi, Martin;

Thanks for your reply!




Maybe this idea is just for chip company internal dram test.

For examle:

dram is not stable on some platforms, uefi bios could not boot into shell environment, 

firmware developer wants to debug the unstable reason with memory test tool.




This version may extend memtest tool's usage scenario.




current memtest tool supports serial output if gop not exists.

so, during pei phase, user could use uart port as memtest ouput device.
















At 2024-07-24 15:50:23, "martinwhitaker" ***@***.***> wrote:

Memtest86+ uses the EFI Boot Services to obtain the physical memory map and uses the EFI Graphics Output Protocol to locate and configure the video frame buffer. Are those services/protocols available in the PEI phase? I suspect not.

Also Memtest86+ overwrites the memory regions used by the UEFI BIOS, so once it has run there is no possibility of returning to the UEFI BIOS except by rebooting. So what do you gain by running it during the PEI phase?
